dan_layer/engine/tests/events.rs

```diff
@@ -1,31 +1,151 @@
+            vec![Instruction::CallFunction {
+                template_address: faucet_template,
+                function: "mint".to_string(),
+                args: args![initial_supply],
+            }],
+            vec![template_test.get_test_proof()],
+        )
+        .unwrap();
+    let faucet_component: ComponentAddress = result.finalize.execution_results[0].decode().unwrap();
+    let faucet_resource = result
+        .finalize
+        .result
+        .expect("Faucet mint failed")
+        .up_iter()
+        .find_map(|(address, _)| address.as_resource_address())
+        .unwrap();
+
+    // Create sender and receiver accounts
+    let (sender_address, sender_proof, _) = template_test.create_owned_account();
+    let (receiver_address, _, _) = template_test.create_owned_account();
+    template_test
+        .execute_and_commit(
+            vec![
+                Instruction::CallMethod {
+                    component_address: faucet_component,
+                    method: "take_free_coins".to_string(),
+                    args: args![],
+                },
+                Instruction::PutLastInstructionOutputOnWorkspace {
+                    key: b"free_coins".to_vec(),
+                },
+                Instruction::CallMethod {
+                    component_address: sender_address,
+                    method: "deposit".to_string(),
+                    args: args![Variable("free_coins")],
+                },
+            ],
+            vec![template_test.get_test_proof()],
+        )
+        .unwrap();
+
+    // transfer some tokens between accounts
+    let amount = Amount(100);
+    let result = template_test
+        .execute_and_commit(
+            vec![
+                Instruction::CallMethod {
+                    component_address: sender_address,
+                    method: "withdraw".to_string(),
+                    args: args![faucet_resource, amount],
+                },
+                Instruction::PutLastInstructionOutputOnWorkspace {
+                    key: b"foo_bucket".to_vec(),
+                },
+                Instruction::CallMethod {
+                    component_address: receiver_address,
+                    method: "deposit".to_string(),
+                    args: args![Variable("foo_bucket")],
+                },
+            ],
+            // Sender proof needed to withdraw
+            vec![sender_proof],
+        )
+        .unwrap();
+
+    assert!(result.finalize.is_accept());
+
+    // a standard event for the withdraw must have been emmitted
+    assert!(result.finalize.events.iter().any(|e| {
+        e.topic() == "std.vault.withdraw".to_owned() &&
+            e.template_address() == ACCOUNT_TEMPLATE_ADDRESS &&
+            e.component_address().unwrap() == sender_address &&
+            *e.payload().get("resource_address").unwrap() == faucet_resource.to_string() &&
+            *e.payload().get("amount").unwrap() == amount.to_string()
+    }));
```

Comment:
Prefer if each of these are separate asserts, will make it easier to tell what exactly is wrong if the test fails

```rust
assert_eq!(e.topic(), "std.vault.withdraw");
// etc
```